### PR TITLE
8.0 Compatible changes

### PR DESCRIPTION
**Pull request by JayVora-SerpentCS**

### CODE REVIEW

File: contract_isp/__init__.py [hunk 1]
```diff
@@ -3,8 +3,9 @@
 ##############################################################################
 #
```
Comment (elicoidal) [inline contract_isp/__init__.py]: Please use simple headers in all py files following OCA templates
https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3

File: contract_isp/__openerp__.py [hunk 2]
```diff
@@ -28,7 +29,8 @@
 Manage service based contracts
```
Comment (elicoidal) [inline contract_isp/__openerp__.py]: README:
Please create a OCA README file:
https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst
Transfer the description in there.
Follow the template for openerp.py: https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py
version: 8.0.1.0.0

File: contract_isp/models/contract.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+            else:
+                date_format = '%Y/%m/%d'
+
+            if line.analytic_line_type == 'r':
+                if mode == 'prorata':
+                    activation_date = date
+                    month_days = calendar.monthrange(activation_date.year,
+                                                     activation_date.month)[1]
+                    used_days = month_days - activation_date.day
+                    ptx = (100 * used_days / month_days) / 100.0
```
Comment (elicoidal) [inline contract_isp/models/contract.py]: why multiply by 100 then divide by 100? multiplication is commutative

File: contract_isp_automatic_invoicing/__openerp__.py [hunk 4]
```diff
@@ -24,12 +25,13 @@
     'name': 'Contract ISP Automatic Invoicing',
```
Comment (elicoidal) [inline contract_isp_automatic_invoicing/__openerp__.py:25]: same remarks as above

File: contract_isp_automatic_invoicing/__openerp__.py [hunk 5]
```diff
@@ -24,12 +25,13 @@
...
     'version': '1.0',
     'category': 'Account',
-    'description': """A module to automatically invoice services based contracts""",
```
Comment (elicoidal) [inline contract_isp_automatic_invoicing/__openerp__.py:27]: this is summary
service-based

File: contract_isp_automatic_invoicing/__openerp__.py [hunk 6]
```diff
@@ -24,12 +25,13 @@
...
-    'author': "Savoir-faire Linux Inc.,Odoo Community Association (OCA)",
+    'description': """A module to automatically invoice services
```
Comment (elicoidal) [inline contract_isp_automatic_invoicing/__openerp__.py:28]: remove

File: contract_isp_invoice/__openerp__.py [hunk 7]
```diff
@@ -36,17 +37,17 @@
 * Refund on contract closing
 * Exception services invoice logic
 """,
-    'author': "Savoir-faire Linux (joao.gama@savoirfairelinux.com),Odoo Community Association (OCA)",
+    'author': "Savoir-faire Linux Inc, Odoo Community Association (OCA), Serpent Consulting Services Pvt. Ltd.",
```
Comment (elicoidal) [inline contract_isp_invoice/__openerp__.py:40]: add contributors in the README file

File: contract_isp_invoice/models/contract.py [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+    @api.model
+    def create(self, data):
+        if self._context is None:
+            pass
+        if self._context.get('original_amount', False) and data.get('amount',
+                                                                    False):
+            if data['amount'] < data['original_amount']:
+                raise Warning(
+                    _('Error'),
+                    _('Amount cannot be less than the invoice amount'))
```
Comment (elicoidal) [inline contract_isp_invoice/models/contract.py:93]: period

File: contract_isp_invoice/models/contract.py [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+        for journal_type, account_ids in journal_types.items():
+            for account in analytic_account_obj.browse(list(account_ids)):
+                # jgama - If there's a parent, invoice the parent
+                partner = account.partner_id.parent_id or account.partner_id
+
+                if (not partner) or not (account.pricelist_id):
+                    raise Warning(
+                        _('Analytic Account Incomplete!'),
+                        _('Contract incomplete. Please fill in the Customer'
+                          'and Pricelist fields.'))
```
Comment (elicoidal) [inline contract_isp_invoice/models/contract.py:388]: s/and/and/or
(maybe better to have 2 checks and 2 messages)

File: contract_isp_invoice/models/contract.py [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+                # set company_id in context, so the correct default journal
+                # will be selected
+                context2['force_company'] = curr_invoice['company_id']
+                # set force_company in context so the correct product
+                # properties are selected (eg. income account)
+                context2['company_id'] = curr_invoice['company_id']
+                last_invoice = invoice_obj.create(curr_invoice)
+                invoices.append(last_invoice.id)
+
+                self._cr.execute("""SELECT product_id, user_id, to_invoice,
```
Comment (elicoidal) [inline contract_isp_invoice/models/contract.py:422]: I am not comfortable with SQL statement.
There is currently no check on company_id which I think might be necessary

File: contract_isp_invoice/models/contract.py [hunk 11]
```diff
@@ -0,0 +1,535 @@
...
+                self._cr.fetchall():
+                    context2.update({'uom': uom})
+
+                    if data.get('product'):
+                        # force product, use its public price
+                        product_id = data['product'][0]
+                        unit_price = self._get_invoice_price(account,
+                                                             product_id,
+                                                             user_id, qty)
+                    #  elif journal_type == 'general' and product_id:
```
Comment (elicoidal) [inline contract_isp_invoice/models/contract.py:444]: remove dead code

File: contract_isp_invoice/models/contract.py [hunk 12]
```diff
@@ -0,0 +1,535 @@
...
+                    #  timesheets, use sale price
+                    #  unit_price = self._get_invoice_price(cr, uid,
+                    #  account, product_id, user_id, qty, context2)
+                    else:
+                        #  expenses, using price from amount field
+                        unit_price = qty and total_price * -1.0 / qty or 0.0
```
Comment (elicoidal) [inline contract_isp_invoice/models/contract.py:450]: is division by zero handled here?

File: contract_isp_invoice/models/contract.py [hunk 13]
```diff
@@ -0,0 +1,535 @@
...
+
+                    factor = invoice_factor_obj.browse(factor_id)
+                    # factor_name = factor.customer_name and line_name +
```
Comment (elicoidal) [inline contract_isp_invoice/models/contract.py:453]: remove dead code

File: contract_isp_invoice/models/contract.py [hunk 14]
```diff
@@ -0,0 +1,535 @@
...
+                    if product:
+                        factor_name = data.get('product_name', '') and \
+                            product.name_get()
+                        if factor.customer_name and data.get('factor_name',
+                                                             False):
+                            factor_name += ' - ' + factor.customer_name
+
+                        general_account = product.property_account_income or \
+                            product.categ_id.property_account_income_categ
+                        if not general_account:
```
Comment (elicoidal) [inline contract_isp_invoice/models/contract.py:479]: checks and warning should be done at the beginning of the method, not in the middle

File: contract_isp_invoice/models/contract.py [hunk 15]
```diff
@@ -0,0 +1,535 @@
...
+                            raise Warning(
+                                _("Configuration Error!"),
+                                _("Please define income account for"
+                                  "product '%s'.") % product.name)
+                        taxes = product.taxes_id or general_account.tax_ids
+                        tax = fiscal_pos_obj.map_tax(taxes)
+                        curr_line.update({
+                            'invoice_line_tax_id': [(6, 0, [tax.id])],
+                            'name': factor_name,
+                            #   'invoice_line_tax_id': [(6, 0, tax.id)],
```
Comment (elicoidal) [inline contract_isp_invoice/models/contract.py:489]: remove

File: contract_isp_invoice/models/contract.py [hunk 16]
```diff
@@ -0,0 +1,535 @@
...
+                            'account_id': general_account.id,
+                        })
+                    #
+                    # Compute for lines
+                    #
+                    self._cr.execute("SELECT * FROM account_analytic_line"
```
Comment (elicoidal) [inline contract_isp_invoice/models/contract.py:495]: same remark as above (about multi-company)

File: contract_isp_package_configurator/__openerp__.py [hunk 17]
```diff
@@ -28,8 +29,10 @@
 Contract ISP Package Configurator
```
Comment (elicoidal) [inline contract_isp_package_configurator/__openerp__.py:29]: README+openerp.py

File: contract_isp_package_configurator/models/company.py [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+from openerp import models, fields
+
+
+class res_company(models.Model):
+    _inherit = 'res.company'
+
+    default_product_category = fields.Many2one('product.category',
+                                               'Default Product Category',
+                                               required=True,
+                                               help='''Set the default'''
```
Comment (elicoidal) [inline contract_isp_package_configurator/models/company.py:32]: You can improve the indentation here

File: contract_isp_package_configurator/models/company.py [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+                                               '''starting point for'''
+                                               ''' the package'''
+                                               '''configuration wizard''')
```
Comment (elicoidal) [inline contract_isp_package_configurator/models/company.py:35]: Help seems unrelated with the field definition.

File: contract_isp_package_configurator/views/res_config_view.xml [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+          <separator string="Package configuration" />
+
+          <group>
+            <label for="id" string="Default options" />
+
+            <div>
+              <div>
+                <label for="default_category" />
+                <field name="default_category" class="oe_inline" />
+                <!-- FIXME domaine de default_category -->
```
Comment (elicoidal) [inline contract_isp_package_configurator/views/res_config_view.xml:18]: easy to fix now

File: product_dependencies/__openerp__.py [hunk 21]
```diff
@@ -4,7 +4,8 @@
 #
```
Comment (elicoidal) [inline product_dependencies/__openerp__.py:4]: same remarks: headers/README/openerp.py

**Comment by JayVora-SerpentCS** [comment]
Thanks, will check soon.

File: contract_isp/wizard/activate_contract_service.py [hunk 22]
```diff
@@ -1,51 +1,38 @@
...
+#    (<http://www.serpentcs.com>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
 
 import datetime
-from openerp import models, fields, api
+from openerp.tools import DEFAULT_SERVER_DATETIME_FORMAT
+from openerp import api, fields, models, _
+
+
+@api.one
```
Comment (joaoalf) [inline contract_isp/wizard/activate_contract_service.py:12]: @JayVora-SerpentCS Better to avoid using `@api.one` since it's deprecated in newer versions of odoo

File: contract_isp/wizard/activate_contract_service.py [hunk 23]
```diff
@@ -1,51 +1,38 @@
...
+def _get_account_id(self):
+    if self._context.get('active_model', '') == 'contract.service':
+        contract_id = self._context.get('active_id')
+        contract_service = self.env['contract.service'].browse(contract_id)
+        return contract_service.account_id.id
+    return None
+
+
+@api.one
```
Comment (joaoalf) [inline contract_isp/wizard/activate_contract_service.py:21]: @JayVora-SerpentCS Better to avoid using `@api.one` since it's deprecated in newer versions of odoo

File: contract_isp/wizard/activate_contract_service.py [hunk 24]
```diff
@@ -59,31 +46,90 @@ def _get_service_id(self):
...
+    deactivation_date = fields.Datetime('Deactivation Date',
+                                        default=fields.datetime.now())
+    account_id = fields.Many2one('account.analytic.account', 'Account',
+                                 default=_get_account_id)
+    service_id = fields.Many2one('contract.service', 'Service',
+                                 default=_get_service_id)
+
+    @api.multi
+    def deactivate(self):
+        context = self._context
```
Comment (joaoalf) [inline contract_isp/wizard/activate_contract_service.py:98]: @JayVora-SerpentCS I don't get why you assign context 3 times here. Why can't you use `obj.with_context(..).action_desactivate(...)` ?

File: contract_isp/wizard/activate_contract_service.py [hunk 25]
```diff
@@ -59,31 +46,90 @@ def _get_service_id(self):
...
                     line.create_analytic_line(mode='manual',
                                               date=activation_date)
-
                     if line.analytic_line_type == 'r':
                         line.create_analytic_line(mode='prorata',
                                                   date=activation_date)
+        return True
+
+    @api.constrains('activation_date')
+    def check_future_date(self):
```
Comment (joaoalf) [inline contract_isp/wizard/activate_contract_service.py:73]: @JayVora-SerpentCS Missing `_`.

File: contract_isp/wizard/activate_contract_service.py [hunk 26]
```diff
@@ -59,31 +46,90 @@ def _get_service_id(self):
...
+                return False
+        return True
+
+
+class contract_service_deactivate(models.TransientModel):
+
+    _name = 'contract.service.deactivate'
+
+    deactivation_date = fields.Datetime('Deactivation Date',
+                                        default=fields.datetime.now())
```
Comment (joaoalf) [inline contract_isp/wizard/activate_contract_service.py:90]: @JayVora-SerpentCS Please use `lambda` in default fields unless it's a constant. Also, `datetime->Datetime`
